### PR TITLE
UnityXR SDK Support

### PR DESCRIPTION
**Pull request by ddugganCRA**
This pull request will add UnityXR support, which is much easier to set up than any of the other SDKs and is a little more flexible.

Usage guide is already posted on the wiki, but a quick note: this is only being actively support for 2019.1 and newer. I included the existing implementation for 2018 from VRTK (and verified that it works), but I will not be maintaining it.

When complete this will resolve: #3

### CODE REVIEW

File: Assets/VIRTUOSO/Scripts/HapticSDK/Haptic Devices/UnityDevice.cs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    /// Author: Dan Duggan (dduggan@cra.com) Updated: September 2019
+    /// </summary>
+    [HapticSystem("Unity", "Left Controller", "BodyCoordinates/LeftHand", typeof(SDK_UnityController), false)]
+    [HapticSystem("Unity", "Right Controller", "BodyCoordinates/RightHand", typeof(SDK_UnityController), true)]
+    public class UnityDevice : HapticDevice
+    {
+#if UNITY_2018_3_OR_NEWER
+        #region Protected Variables
+        // Indicates which hand this object corresponds to
+        protected XRNode nodeType;
```
Comment (awollocko) [inline Assets/VIRTUOSO/Scripts/HapticSDK/Haptic Devices/UnityDevice.cs]: indicate "hand" in the variable name?

File: Assets/VIRTUOSO/Scripts/HapticSDK/Haptic Devices/UnityDevice.cs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+            }
+            else
+            {
+                Debug.LogWarningFormat("Unity XR Haptic Device {0} failed to cancel: device is invalid", nodeType.ToString());
+            }
+        }
+
+        protected override void StartHaptics(HumanBodyBones bodyPart, BodyCoordinateHit hitLocation, float intensity)
+        {
+#if UNITY_2019_1_OR_NEWER
```
Comment (awollocko) [inline Assets/VIRTUOSO/Scripts/HapticSDK/Haptic Devices/UnityDevice.cs]: I'd create a helper function that returns a boolean - something like "IsHapticDeviceValid" and do the platform checks within that.

File: Assets/VIRTUOSO/Scripts/HapticSDK/Haptic Devices/UnityDevice.cs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        }
+
+        /// <summary>
+        /// This override is used to determine the correct XRNode type for this device.
+        /// </summary>
+        public override void ApplyDefaultData(HapticSystemAttribute hapticSystemInfo)
+        {
+            base.ApplyDefaultData(hapticSystemInfo);
+
+            nodeType = (bool)hapticSystemInfo.AdditionalData[0] ? XRNode.RightHand : XRNode.LeftHand;
```
Comment (awollocko) [inline Assets/VIRTUOSO/Scripts/HapticSDK/Haptic Devices/UnityDevice.cs]: comment

File: Assets/VIRTUOSO/Scripts/HapticSDK/Haptic Devices/UnityDevice.cs [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        protected override void Instance_LoadedSetupChanged(VRTK_SDKManager sender, VRTK_SDKManager.LoadedSetupChangeEventArgs e)
+        {
+            if (e.currentSetup.controllerSDKInfo.type == hapticSystemInfo.ConnectedSDKType)
+            {
+                hapticDevice = InputDevices.GetDeviceAtXRNode(nodeType);
+            }
+        }
+        #endregion
+#else
+        protected override void CancelHaptics()
```
Comment (awollocko) [inline Assets/VIRTUOSO/Scripts/HapticSDK/Haptic Devices/UnityDevice.cs:105]: can you then remove the overrides or is there value in having them here?
If purposeful, provide a leading comment.

File: Assets/VIRTUOSO/VRTK/Source/SDK/Unity/SDK_UnityController.cs [hunk 5]
```diff
@@ -138,15 +210,101 @@ Axis Codes
...
+#endif
         }
 
+#if UNITY_2019_1_OR_NEWER
+        /// <summary>
+        /// Uses Unity XR input to determine the status of each button on the controller.
+        /// </summary>
+        protected void ProcessUnityXRInput(Dictionary<ButtonTypes,UnityXRInputState> handState, InputDevice currentController)
+        {
+            if(currentController.isValid != true)
```
Comment (awollocko) [inline Assets/VIRTUOSO/VRTK/Source/SDK/Unity/SDK_UnityController.cs:231]: unsure but wanted to ask based on the other files -- does this need a UNITY_VERSION_CHECK?

File: Assets/VIRTUOSO/VRTK/Source/SDK/Unity/SDK_UnityController.cs [hunk 6]
```diff
@@ -138,15 +210,101 @@ Axis Codes
...
+            // get current button values
+            foreach (ButtonTypes buttonType in handState.Keys)
+            {
+                bool lastFrameTouchValue = handState[buttonType].buttonState[ButtonPressTypes.Touch];
+                bool lastFramePressValue = handState[buttonType].buttonState[ButtonPressTypes.Press];
+
+                bool currentFrameTouchValue = false;
+                bool currentFramePressValue = false;
+                bool buttonValueOut;
+                float floatValueOut = 0.0f;
```
Comment (awollocko) [inline Assets/VIRTUOSO/VRTK/Source/SDK/Unity/SDK_UnityController.cs]: no need for "float" in the variable name. Also, specific on what the value is out?

File: Assets/VIRTUOSO/VRTK/Source/SDK/Unity/SDK_UnityController.cs [hunk 7]
```diff
@@ -157,7 +315,12 @@ public override void ProcessFixedUpdate(VRTK_ControllerReference controllerRefer
         public override ControllerType GetCurrentControllerType(VRTK_ControllerReference controllerReference = null)
         {
             SetTrackedControllerCaches();
+#if !UNITY_2019_1_OR_NEWER
             return cachedControllerType;
+#else
+            // not used post-2019
```
Comment (awollocko) [inline Assets/VIRTUOSO/VRTK/Source/SDK/Unity/SDK_UnityController.cs:319]: pre-2019

File: Assets/VIRTUOSO/VRTK/Source/SDK/Unity/SDK_UnityController.cs [hunk 8]
```diff
@@ -696,6 +907,14 @@ protected virtual void SetTrackedControllerCaches(bool forceRefresh = false)
                         cachedLeftTracker = cachedLeftController.GetComponent<SDK_UnityControllerTracker>();
                         cachedLeftVelocityEstimator = cachedLeftController.GetComponent<VRTK_VelocityEstimator>();
                         SetControllerButtons(ControllerHand.Left);
+#if UNITY_2019_1_OR_NEWER
+                        List<InputDevice> leftHandDevices = new List<InputDevice>();
+                        InputDevices.GetDevicesAtXRNode(XRNode.LeftHand, leftHandDevices);
+                        if(leftHandDevices.Count > 0)
```
Comment (awollocko) [inline Assets/VIRTUOSO/VRTK/Source/SDK/Unity/SDK_UnityController.cs:913]: wouldn't a leftHandDevices[0] != null work too?

File: Assets/VIRTUOSO/VRTK/Source/SDK/Unity/SDK_UnityControllerTracker.cs [hunk 9]
```diff
@@ -21,6 +25,7 @@ public class SDK_UnityControllerTracker : MonoBehaviour
         public XRNode nodeType;
         [Tooltip("The unique index to assign to the controller.")]
         public uint index;
+#if !UNITY_2019_1_OR_NEWER
         [Tooltip("The Unity Input name for the trigger axis.")]
         public string triggerAxisName = "";
```
Comment (awollocko) [inline Assets/VIRTUOSO/VRTK/Source/SDK/Unity/SDK_UnityControllerTracker.cs]: Should have a default value? Pick a platform, most likely?

**Review by awollocko — CHANGES_REQUESTED** [review]

File: Assets/VIRTUOSO/Scripts/HapticSDK/Haptic Devices/UnityDevice.cs (hunk 2 above)
Comment (ddugganCRA) [reply Assets/VIRTUOSO/Scripts/HapticSDK/Haptic Devices/UnityDevice.cs]: Good call, the defines were getting out of hand

File: Assets/VIRTUOSO/Scripts/HapticSDK/Haptic Devices/UnityDevice.cs (hunk 4 above)
Comment (ddugganCRA) [reply Assets/VIRTUOSO/Scripts/HapticSDK/Haptic Devices/UnityDevice.cs:105]: Added comment to clarify, but the base methods are abstract so need an implementation

File: Assets/VIRTUOSO/VRTK/Source/SDK/Unity/SDK_UnityController.cs (hunk 5 above)
Comment (ddugganCRA) [reply Assets/VIRTUOSO/VRTK/Source/SDK/Unity/SDK_UnityController.cs:231]: Yes, this method is surrounded by one as the entire method and parameters depend on 2019+ changes

File: Assets/VIRTUOSO/VRTK/Source/SDK/Unity/SDK_UnityController.cs (hunk 7 above)
Comment (ddugganCRA) [reply Assets/VIRTUOSO/VRTK/Source/SDK/Unity/SDK_UnityController.cs:319]: rearranged this as the define setup was a little confusing in this spot

File: Assets/VIRTUOSO/VRTK/Source/SDK/Unity/SDK_UnityController.cs (hunk 8 above)
Comment (ddugganCRA) [reply Assets/VIRTUOSO/VRTK/Source/SDK/Unity/SDK_UnityController.cs:913]: No, the list can be empty so the leftHandDevices[0]  call would result in an exception

File: Assets/VIRTUOSO/VRTK/Source/SDK/Unity/SDK_UnityControllerTracker.cs (hunk 9 above)
Comment (ddugganCRA) [reply Assets/VIRTUOSO/VRTK/Source/SDK/Unity/SDK_UnityControllerTracker.cs]: The axis is hand dependent, so I don't want to assign a default. However, reviewing this code myself (it was from the original VRTK stuff) I don't see it actually used anywhere. I'm going to double check with a 2017 build and if it's not used I'm removing it.

Comment (ddugganCRA) [reply Assets/VIRTUOSO/VRTK/Source/SDK/Unity/SDK_UnityControllerTracker.cs]: It is used in pre-2019 versions. The value is defined in the prefab, so I am changing it to use the Left Hand versions by default and will add a comment to reflect that.

File: Assets/VIRTUOSO/Prefabs/[UnityBase_CameraRig].prefab [hunk 10]
```diff
@@ -285,11 +285,11 @@ Camera:
   m_TargetEye: 3
   m_HDR: 0
   m_AllowMSAA: 1
+  m_AllowDynamicResolution: 0
   m_ForceIntoRT: 0
   m_OcclusionCulling: 1
   m_StereoConvergence: 10
   m_StereoSeparation: 0.022
-  m_StereoMirrorMode: 0
 --- !u!54 &54108975209875660
```
Comment (VR-Nic) [inline Assets/VIRTUOSO/Prefabs/[UnityBase_CameraRig].prefab:293]: When I imported this prefab, the 'HandNodeType' variable on the SDK_UnityControllerTracker script were not set to the proper hands and set to the default 'LeftEye'.

**Review by VR-Nic — CHANGES_REQUESTED** [review]
I tested this branch using Unity 2019.3.0a7, an HTC Vive, and SteamVR 1.7.15.

The UnitySDK works and I was able to set up basic interactable objects. However, there were some issues I encountered that should be noted. First off, if SDKManager has 'AutoManager VR Settings' enabled when the user starts their project, then they need to move or delete the 'None' option that it adds to the VR list otherwise they will run into issues (add this in the Wiki). Second, in the Wiki, it should mention that models are not set up for the hand and the user will need to provide a visual representation for the controllers (and where to place them). Last, in the Wiki for the UnitySDK, it says to create an Empty GameObject under the SDKManager, but all SDKs are placed under [VRTK_SDKSetups] (Change Step 5 in Wiki).

Other than that, there is only 1 issue with a Prefab that I ran into (specified in the file). Very minor changes from me. Good work Dan.

**Comment by VR-Nic** [comment]
One additional thing, and this is for all future PRs and such, but the build list of scenes should be cleared if we're not including the scenes. Otherwise, Unity only specifies that they are deleted but it does not let you select and remove them.
UnityBuildList

File: Assets/VIRTUOSO/Prefabs/[UnityBase_CameraRig].prefab (hunk 10 above)
Comment (ddugganCRA) [reply Assets/VIRTUOSO/Prefabs/[UnityBase_CameraRig].prefab:293]: Fixed

**Comment by ddugganCRA** [comment]
@VR-Nic @awollocko I've made the requested changes, please re-review when you get a chance.

**Review by VR-Nic — APPROVED** [review]

**Pull request merged**